apps/snfoundry/contracts/src/Marketplace.cairo

```diff
@@ -0,0 +1,429 @@
+    struct PaymentSeller {
+        token_ids: Span<u256>,
+        seller: ContractAddress,
+        payment: u256,
+    }
+
+    ///
+    /// Constructor.
+    /// # Arguments
+    /// * `cofi_collection_address` - The address of the CofiCollection contract
+    /// * `admin` - The address of the admin role
+    /// * `market_fee` - The fee that the marketplace will take from the sales
+    /// * `base_uri` - The base uri for the NFTs metadata. Should contain `{id}` so that metadata gets
+    ///    replace per each token id. Example: https://example.com/metadata/{id}.json
+    ///
+    #[constructor]
+    fn constructor(
+        ref self: ContractState,
+        cofi_collection_address: ContractAddress,
+        admin: ContractAddress,
+        market_fee: u256,
+        base_uri: ByteArray,
+    ) {
+        self.erc1155_receiver.initializer();
+        self.accesscontrol.initializer();
+        self.accesscontrol._grant_role(DEFAULT_ADMIN_ROLE, admin);
+        self.cofi_collection_address.write(cofi_collection_address);
+        self.market_fee.write(market_fee);
+        self.current_token_id.write(1);
+
+        ICofiCollectionDispatcher {contract_address:  self.cofi_collection_address.read()}.set_base_uri(base_uri);
+    }
+
+    #[abi(embed_v0)]
+    impl MarketplaceImpl of super::IMarketplace<ContractState> {
+        fn assign_seller_role(ref self: ContractState, assignee: ContractAddress) {
+            self.accesscontrol.assert_only_role(DEFAULT_ADMIN_ROLE);
+            self.accesscontrol._grant_role(PRODUCER, assignee);
+        }
+    
+        fn assign_consumer_role(ref self: ContractState, assignee: ContractAddress) {
+            self.accesscontrol.assert_only_role(DEFAULT_ADMIN_ROLE);
+            self.accesscontrol._grant_role(CONSUMER, assignee);
+        }
+
+        fn assign_admin_role(ref self: ContractState, assignee: ContractAddress) {
+            self.accesscontrol.assert_only_role(DEFAULT_ADMIN_ROLE);
+            self.accesscontrol._grant_role(DEFAULT_ADMIN_ROLE, assignee);
+        } 
+    
+        fn buy_product(ref self: ContractState, token_id: u256, token_amount: u256) {
+            let stock = self.listed_product_stock.read(token_id);
+            assert(stock >= token_amount, 'Not enough stock');
+
+            let buyer = get_caller_address();
+            let contract_address = get_contract_address();
+            let strk_token_dispatcher = IERC20Dispatcher {
+                contract_address: contract_address_const::<STRK_TOKEN_ADDRESS>()
+            };
+            // Get payment from buyer
+            let mut producer_fee = self.listed_product_price.read(token_id) * token_amount;
+            let mut total_price = producer_fee + self.calculate_fee(producer_fee, self.market_fee.read());
+            assert(strk_token_dispatcher.balance_of(get_caller_address()) >= total_price, 'insufficient funds');
+            strk_token_dispatcher.transfer_from(buyer, contract_address, total_price);
```

Comment:
I'm not totally sure if total_price should be expressed in strk or in fri... in here it is being sended as fri to avoid decimals

Reply:
I asked Glihm and he confirm this is in fri so we have no issues then... the issue would have been if this was in STRK because we can't handle floating point easily in cairo.